inst/scripts/SEL3D.R

```diff
@@ -0,0 +1,251 @@
+
+    #' Details.
+    #'
+    #' @param targetSeq input sequence to be queried, must be cDNA of a length divisible by 3.
+    #' @param startPos start position of the sequence, set to 1 by default, altered in recursive steps
+    #' @param endPos end position of sequence, set to 0, altered to aa length in recursive steps
+    #' @param convert bool to convert DNA sequences to amino acids, True by default
+    #' @param silent bool, set to True to turn off all messages.
+    #' @return MT object - dataframe with headers queryid, subjectids, identity, positives,
+    #'                     alignment length, mismatches, gapopens, q.start, q.end, s.start, s.end,
+    #'                     evalue, bitscore, mlog.evalue, pdb.id
+    #'
+    #'
+    #'
+    #' @examples
+    #' >getPdbIds("AAGTCT")
+    #' >sequence = "ACTGTGACTGTAAATGTT"
+    #' >getPdbIds(sequence)
+    #' >getPdbIds("MGEVST", convert=FALSE)
+  #check for valid targetSeq
+  if(targetSeq =="") {
+    stop(paste0("Error: empty targetSeq"))
+  }
+  if(convert && !nchar(targetSeq) %% 3 == 0) {
+    stop(paste0("Error: targetSeq's length needs to be divisible by 3"))
+  }
+  #we want to blast sequences of amino acid length >50 so the input cDNA length should be >150
+  if(nchar(targetSeq) < 150) {
+    stop(paste0("Error: targetSeq's length must be >150"))
+  }
+  if(convert) {
+    if(grepl("[^AGTC]", targetSeq)) {
+      stop(paste0("Error: targetSeq contains invalid characters"))}
+  }
+
+  #convert sequence into AA
+  if(convert){
+    #translate needs a vector of single chars, not a string
+    AAseq <- paste(seqinr::translate(seqinr::s2c(targetSeq)), collapse='')
+    endPos <- nchar(AAseq)
+  }else{ AAseq <- targetSeq }
+
+  if(silent) {
+    suppressMessages(capture.output(blastOutput <- bio3d::blast.pdb(AAseq)))
+  } else {
+    blastOutput <- bio3d::blast.pdb(AAseq)
+  }
+
+
+  #First we drop any match with ID < 40 to save runtime since we
+  #don't want to include any of these results anyway (base case)
+  blastOutput <- blastOutput$hit.tbl
+  numEntriesBeforeIdentityDrop <- nrow(blastOutput)
+  blastOutput <- blastOutput[blastOutput$identity > 40, ]
+  numEntriesAfterIdentityDrop <- nrow(blastOutput)
+
+  if(!silent) {
+    message(sprintf("dropped %d entries because of identity <= 40",
+                    numEntriesBeforeIdentityDrop - numEntriesAfterIdentityDrop))
+  }
+
+  if(nrow(blastOutput) == 0){
+    return(NULL)
+  }
+
+  blastOutput$tax <- NA
+  blastOutput$info <- NA
+
+  finalOutput <- data.frame()
+  finalOutputRange <- c()
+  #Pointer to keep track of position to insert alignment ranges into final output
+  finalOutputRangePointer <- 1
+
+  for(entry in 1:nrow(blastOutput)){
+    #We're looking up taxonomy and the description of the pdbID
+    #from a different server which uses slightly different formatting
+    #Reformat PDB ID as output by BLAST to separate ID from chain with a "."
+    RCSBid <- gsub("_", ".", blastOutput$pdb.id[entry], ignore.case = TRUE)
```

Comment:
It's not about which taxonomy we're blasting against it's that PDB and RCSB use different formatting.